oqsprov/oqsprov_keys.c

```diff
@@ -157,26 +260,115 @@ int oqsx_key_fromdata(OQSX_KEY *key, const OSSL_PARAM params[], int include_priv
+        // Workaround for bug in OSSL3:
+        // i2d_PrivateKey returns incorrect size. We ignore it because the encoded data seems correct.
+        // One could instead point to the already allocated private key once a bug for this is fixed
+        // See: https://github.com/openssl/openssl/issues/14655
+        unsigned char *pkey_enc = NULL;
+        i2d_PrivateKey(pkey, &pkey_enc);
+        ON_ERR_SET_GOTO(!pkey_enc, ret, -1, errhyb);
+
+        privkeykexlen = ctx->kex_info->kex_length_private_key;
+        memcpy(privkey, pkey_enc, ctx->kex_info->kex_length_private_key);
+        OPENSSL_clear_free(pkey_enc, privkeykexlen);
+    }
+
+    errhyb:
+    EVP_PKEY_CTX_free(kgctx);
+    EVP_PKEY_free(pkey);
+    OPENSSL_free(pubkeykex_encoded);
+
+    return ret;
+}
+
 int oqsx_key_gen(OQSX_KEY *key)
 {
-    if (key->privkey == NULL || key->pubkey == NULL) oqsx_key_allocate_keymaterial(key);
+    int ret = 0;
+
+    if (key->privkey == NULL || key->pubkey == NULL) {
+        ret = oqsx_key_allocate_keymaterial(key);
+        ON_ERR_GOTO(ret, err);
+    }
 
-    if (key->iskem)
-        return OQS_KEM_keypair(key->key.k, key->pubkey, key->privkey);
-    else
-        return OQS_SIG_keypair(key->key.s, key->pubkey, key->privkey);
+    if (key->keytype == KEY_TYPE_KEM) {
+        key->privkey_ptr[0] = key->privkey;
+        key->pubkey_ptr[0] = key->pubkey;
+        ret = oqsx_key_gen_oqs_kem(key->oqsx_provider_ctx.oqsx_qs_ctx.kem, key->pubkey_ptr[0], key->privkey_ptr[0]);
+        ON_ERR_GOTO(ret, err);
+    } else if (key->keytype == KEY_TYPE_ECP_HYB_KEM || key->keytype == KEY_TYPE_ECX_HYB_KEM) {
+        key->privkey_ptr[0] = key->privkey;
+        key->pubkey_ptr[0] = key->pubkey;
+        key->privkey_ptr[1] = key->privkey + key->oqsx_provider_ctx.oqsx_evp_ctx->kex_info->kex_length_private_key;
+        key->pubkey_ptr[1] = key->pubkey + key->oqsx_provider_ctx.oqsx_evp_ctx->kex_info->kex_length_public_key;
+        ret = oqsx_key_gen_evp_kex(key->oqsx_provider_ctx.oqsx_evp_ctx, key->pubkey_ptr[0], key->privkey_ptr[0]);
+        ON_ERR_GOTO(ret, err);
+        ret = oqsx_key_gen_oqs_kem(key->oqsx_provider_ctx.oqsx_qs_ctx.kem, key->pubkey_ptr[1], key->privkey_ptr[1]);
+        ON_ERR_GOTO(ret, err);
+    } else if (key->keytype == KEY_TYPE_SIG) {
+        key->privkey_ptr[0] = key->privkey;
+        key->pubkey_ptr[0] = key->pubkey;
+        ret = oqsx_key_gen_oqs_sig(key->oqsx_provider_ctx.oqsx_qs_ctx.sig, key->pubkey, key->privkey);
+        ON_ERR_GOTO(ret, err);
+    } else {
+        ret = 1;
+    }
+    err:
+    return ret;
 }
 
 int oqsx_key_parambits(OQSX_KEY *key) {
-    if (key->iskem) 
-        return 128+(key->key.k->claimed_nist_level-1)/2*64;
-    return 128+(key->key.s->claimed_nist_level-1)/2*64;
+    if (key->keytype == KEY_TYPE_KEM)
+        return 128+(key->oqsx_provider_ctx.oqsx_qs_ctx.kem->claimed_nist_level-1)/2*64;
+    else if (key->keytype == KEY_TYPE_ECP_HYB_KEM || key->keytype == KEY_TYPE_ECX_HYB_KEM)
```

Comment:
Can the KEM-KEYTYPES not be handled by the same code? To me this is an indication that the "logical merge"/refactoring was successful.

Reply:
The first reason for the distinction is that the init-logic of ECP and ECX curves is still different. The second reason for the two key types is to give an indication to the code, which EC-family should be used for strength-matching.
At least the first point would be nice to be addressed by revisiting the use of new provider API.

Reply:
> revisiting the use of new provider API

what do you mean by this? Are we not running on upstream master? How can there be something even more new?

Reply:
Yes. What I meant was "revisit how to fully adopt the OSSL3 API" (and potential gaps).